### PR TITLE
pylint_plugins: add forbidden import checker

### PR DESCRIPTION
**Pull request by HonzaCholasta**
Add new pylint AST checker plugin which implements a check for imports
forbidden in IPA. Which imports are forbidden is configurable in pylintrc.

Provide default forbidden import configuration and disable the check for
existing forbidden imports in our code base.

Supersedes @MartinBasti's PR #462.

### CODE REVIEW

File: pylint_plugins.py [hunk 1]
```diff
@@ -255,3 +259,79 @@ def fix_ipa_classes(cls):
...
 
 MANAGER.register_transform(scoped_nodes.Class, fix_ipa_classes)
+
+
+class IPAChecker(BaseChecker):
+    __implements__ = IAstroidChecker
+
+    name = 'ipa'
+    msgs = {
+        'E9901': (
```
Comment (MartinBasti) [inline pylint_plugins.py]: Could we use Warning instead of error?

File: pylint_plugins.py [hunk 2]
```diff
@@ -255,3 +259,79 @@ def fix_ipa_classes(cls):
...
+            'Forbidden import %s',
```
Comment (MartinBasti) [inline pylint_plugins.py]: Maybe would be nice to write whihc rule has been applied `Forbidden import %s (%s should not import %s)`

File: pylint_plugins.py (hunk 1 above)
Comment (HonzaCholasta) [inline pylint_plugins.py]: Sure.

File: pylint_plugins.py (hunk 2 above)
Comment (HonzaCholasta) [inline pylint_plugins.py]: I agree.

**Comment by MartinBasti** [comment]
> Can you turn module matching into a regular expression? We need bit more advanced checks, e.g. ipalib should not import from ipaplatform except for modules in ipalib.install.

How can be the issue mentioned by @tiran solved in this PR? should regexp be used or allow rules added?

**Comment by HonzaCholasta** [comment]
@MartinBasti, this issue is already solved in the PR without using regular expressions. See `pylintrc` for example.

**Comment by MartinBasti** [comment]
Ok, this will not work if ipaclient/submodule allows to import any module, but seems OK for me now, can be improved when needed

**Comment by HonzaCholasta** [comment]
I don't know what you mean, could you give me an example?

**Comment by MartinBasti** [comment]
In this case:
```
   ipaclient/:ipaclient.install:ipalib.install:ipaplatform:ipaserver,
   ipaclient/install/:ipaserver,
```

`ipaclient/install` allows all import everything but `ipaserver`, but I cannot currently specify a rule that allows `ipaclient/install` import everything (with `ipaserver`)

But as I said this is a corner case, should be done when needed

**Comment by HonzaCholasta** [comment]
You can, using:
```
    ipaclient/install/
```

**Comment by MartinBasti** [comment]
Awesome then

**Comment by HonzaCholasta** [comment]
The format could be nicer though - suggestions are welcome.

**Comment by martbab** [comment]
@MartinBasti any progress in reviewing this PR?

**Comment by HonzaCholasta** [comment]
@martbab, I haven't incorporated @MartinBasti's suggestions in yet.

**Comment by HonzaCholasta** [comment]
I have now incorporated the suggestions.

**Comment by MartinBasti** [comment]
master:

* 5d489ac5604ca959cfe439c0594b8739073f3cea pylint_plugins: add forbidden import checker

**Pull request closed without merging**